instrumentation/cassandra/cassandra-3.0/src/test/groovy/CassandraClientTest.groovy

```diff
@@ -13,39 +13,45 @@ import io.opentelemetry.instrumentation.test.AgentTestRunner
 import io.opentelemetry.instrumentation.test.asserts.TraceAssert
 import io.opentelemetry.sdk.trace.data.SpanData
 import io.opentelemetry.trace.attributes.SemanticAttributes
+import java.time.Duration
 import java.util.concurrent.Executors
 import java.util.concurrent.atomic.AtomicBoolean
-import org.cassandraunit.utils.EmbeddedCassandraServerHelper
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.testcontainers.containers.GenericContainer
+import org.testcontainers.containers.output.Slf4jLogConsumer
 import spock.lang.Shared
 
 class CassandraClientTest extends AgentTestRunner {
+  private static final Logger log = LoggerFactory.getLogger(CassandraClientTest)
 
   @Shared
-  Cluster cluster
+  def executor = Executors.newCachedThreadPool()
 
   @Shared
-  def executor = Executors.newCachedThreadPool()
+  GenericContainer cassandra
+  @Shared
+  int cassandraPort
+  @Shared
+  Cluster cluster
 
   def setupSpec() {
-    /*
-     This timeout seems excessive but we've seen tests fail with timeout of 40s.
-     TODO: if we continue to see failures we may want to consider using 'real' Cassandra
-     started in container like we do for memcached. Note: this will complicate things because
-     tests would have to assume they run under shared Cassandra and act accordingly.
-      */
-    EmbeddedCassandraServerHelper.startEmbeddedCassandra(EmbeddedCassandraServerHelper.CASSANDRA_RNDPORT_YML_FILE, 120000L)
-
-    cluster = EmbeddedCassandraServerHelper.getCluster()
-
-    /*
-    Looks like sometimes our requests fail because Cassandra takes to long to respond,
-    Increase this timeout as well to try to cope with this.
-     */
-    cluster.getConfiguration().getSocketOptions().setReadTimeoutMillis(120000)
+    cassandra = new GenericContainer("cassandra:3")
+      .withExposedPorts(9042)
+      .withLogConsumer(new Slf4jLogConsumer(log))
+      .withStartupTimeout(Duration.ofSeconds(120))
+    cassandra.start()
```

Comment:
Should I change Cassandra 4.0 tests to use testcontainers as well? WDYT?

Reply:
Yes, please. Although in the future I personally would prefer to have these logically separate changes in separate PRs :)

Reply:
I agree, I probably should have split this PR... but once I started working on this there were so many things that needed to be fixed/improved that I went with the flow and put them all here 😅